### PR TITLE
feat: Image scanning in POST Ci script

### PR DESCRIPTION
**Pull request by Ashish-devtron**
# Description

Plugin which can be configured in Pre/Post Steps to run Image scanning.

Fixes https://github.com/devtron-labs/devtron/issues/4022

## How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

## Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR requires documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code.
* [ ] I have commented my code, particularly in hard-to-understand areas.
* [ ] I have tested it for all user roles.
* [ ] I have added all the required unit/api test cases.

## Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: ciStages.go [hunk 1]
```diff
@@ -73,12 +73,16 @@ func runCIStages(ciCdRequest *helper.CiCdTriggerEvent) (artifactUploaded bool, e
 	start := time.Now()
 	metrics.TotalStartTime = start
 	artifactUploaded = false
+
+	// change the current working directory to '/'
 	err = os.Chdir("/")
```
Comment (kartik-579) [inline ciStages.go:78]: create constant

File: ciStages.go [hunk 2]
```diff
@@ -281,6 +223,119 @@ func runCIStages(ciCdRequest *helper.CiCdTriggerEvent) (artifactUploaded bool, e
...
+		imageRetryCountValue := ciCdRequest.CommonWorkflowRequest.ImageRetryCount
+		imageRetryIntervalValue := ciCdRequest.CommonWorkflowRequest.ImageRetryInterval
+		for i := 0; i < imageRetryCountValue+1; i++ {
+			if i != 0 {
+				time.Sleep(time.Duration(imageRetryIntervalValue) * time.Second)
+			}
+			err = helper.PushArtifact(dest)
+			if err == nil {
+				break
+			}
```
Comment (kartik-579) [inline ciStages.go:299]: can be extracted out

File: ciStages.go [hunk 3]
```diff
@@ -281,6 +223,119 @@ func runCIStages(ciCdRequest *helper.CiCdTriggerEvent) (artifactUploaded bool, e
...
+			util.LogStage("Running POST-CI steps which are enabled to RUN even on CI FAIL")
+			// build success will always be false
+			scriptEnvs[util.ENV_VARIABLE_BUILD_SUCCESS] = "false"
+			// run post artifact processing
+			RunCiCdSteps(STEP_TYPE_POST, postCiStepsToTriggerOnCiFail, refStageMap, scriptEnvs, preCiStageOutVariable)
+		}
+		// code-block ends
+		err = sendFailureNotification(string(Build), ciCdRequest.CommonWorkflowRequest, "", "", metrics, artifactUploaded, err)
+	}
+	log.Println(util.DEVTRON, " /Build")
```
Comment (kartik-579) [inline ciStages.go:275]: improve logs

**Review by kartik-579 — CHANGES_REQUESTED** [review]

File: ciStages.go [hunk 4]
```diff
@@ -139,101 +145,45 @@ func runCIStages(ciCdRequest *helper.CiCdTriggerEvent) (artifactUploaded bool, e
...
-			log.Println("error in getting results", "err", err.Error())
+			return artifactUploaded, err
 		}
 	}
-	metrics.PreCiDuration = preCiDuration
-	var dest string
+
+	var digest string
+
 	if !buildSkipEnabled {
```
Comment (kartik-579) [inline ciStages.go:169]: same condition repeated twice

File: ciStages.go [hunk 5]
```diff
@@ -329,11 +373,31 @@ func makeDockerfile(config *helper.DockerBuildConfig, checkoutPath string) error
...
+	for i := 0; i < imageRetryCountValue+1; i++ {
+		if i != 0 {
+			time.Sleep(time.Duration(imageRetryIntervalValue) * time.Second)
+		}
+		err = helper.PushArtifact(dest)
+		if err == nil {
+			break
+		}
+	}
+	if err != nil {
```
Comment (kartik-579) [inline ciStages.go:398]: we are only handling the error found on last iteration, atleast log the errors found in previous iterations

File: ciStages.go [hunk 6]
```diff
@@ -329,11 +373,31 @@ func makeDockerfile(config *helper.DockerBuildConfig, checkoutPath string) error
...
 	}
-	return artifactUploaded, &helper.CiStageError{Err: err}
+	return &helper.CiStageError{Err: err}
+}
+
+func pushArtifact(ciCdRequest *helper.CiCdTriggerEvent, dest string, digest string, metrics *helper.CIMetrics, artifactUploaded bool) error {
+	imageRetryCountValue := ciCdRequest.CommonWorkflowRequest.ImageRetryCount
+	imageRetryIntervalValue := ciCdRequest.CommonWorkflowRequest.ImageRetryInterval
+	var err error
+	for i := 0; i < imageRetryCountValue+1; i++ {
```
Comment (kartik-579) [inline ciStages.go:389]: why imageRetryCountValue+1 ? we are starting the loop from 0

**Review by kartik-579 — CHANGES_REQUESTED** [review]

**Review by kartik-579 — APPROVED** [review]

File: ciStages.go [hunk 7]
```diff
@@ -282,6 +215,120 @@ func runCIStages(ciCdRequest *helper.CiCdTriggerEvent) (artifactUploaded bool, e
...
+	preCiStageOutVariable, step, err := RunCiCdSteps(STEP_TYPE_PRE, ciCdRequest.CommonWorkflowRequest.PreCiSteps, refStageMap, scriptEnvs, nil)
+	preCiDuration := time.Since(start).Seconds()
+	if err != nil {
+		log.Println("error in running pre Ci Steps", "err", err)
+		err = sendFailureNotification(string(PreCi)+step.Name, ciCdRequest.CommonWorkflowRequest, "", "", *metrics, artifactUploaded, err)
+		return nil, nil, err
+	}
+	// considering pull images from Container repo Plugin in Pre ci steps only.
+	// making it non-blocking if results are not available (in case of err)
+	resultsFromPlugin, err1 := extractOutResultsIfExists()
```
Comment (kartik-579) [inline ciStages.go:237]: use err

File: ciStages.go [hunk 8]
```diff
@@ -140,101 +146,35 @@ func runCIStages(ciCdRequest *helper.CiCdTriggerEvent) (artifactUploaded bool, e
...
 	var postCiDuration float64
 	start = time.Now()
 	metrics.PostCiStartTime = start
 	if len(ciCdRequest.CommonWorkflowRequest.PostCiSteps) > 0 {
-		util.LogStage("running POST-CI steps")
-		// sending build success as true always as post-ci triggers only if ci gets success
-		scriptEnvs[util.ENV_VARIABLE_BUILD_SUCCESS] = "true"
-		// run post artifact processing
-		_, step, err = RunCiCdSteps(STEP_TYPE_POST, ciCdRequest.CommonWorkflowRequest.PostCiSteps, refStageMap, scriptEnvs, preeCiStageOutVariable)
+		err = runPostCiSteps(ciCdRequest, scriptEnvs, refStageMap, preCiStageOutVariable, metrics, artifactUploaded, dest, digest)
```
Comment (kartik-579) [inline ciStages.go:171]: we are passing metrics by value here and above, are they going to be manipulated separately? if not please fix

**Review by kartik-579 — CHANGES_REQUESTED** [review]

**Review by kartik-579 — APPROVED** [review]

**Review by kripanshdevtron — APPROVED** [review]

**Pull request merged**